### PR TITLE
Enable Batch Norm Fusion

### PR DESCRIPTION
**Pull request by rdspring1**
- Add native_batch_norm_backward to parser.cpp
- Enable batch_norm in symbolic_script.cpp

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 1]
```diff
@@ -199,6 +199,52 @@ class NaiveTypePropagator {
...
+        auto out_type = node->input(0)->type()->cast<TensorType>();
+        node->output(0)->setType(out_type);
+
+        auto mean_rstd_type = TensorType::create(
+            *out_type->scalarType(),
+            *out_type->device(),
+            *out_type->dim(),
+            out_type->requires_grad());
+
+        node->output(1)->setType(mean_rstd_type);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp:212]: The dimension of 2nd & 3rd output be 1 instead.

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 2]
```diff
@@ -199,6 +199,52 @@ class NaiveTypePropagator {
...
+        }
+
+        if (output_mask[1]) {
+          auto weight_type = node->input(2)->type()->cast<TensorType>();
+          node->output(1)->setType(weight_type);
+        }
+
+        if (output_mask[2]) {
+          auto weight_type = node->input(2)->type()->cast<TensorType>();
+          auto bias_type = TensorType::create(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp:239]: why do we want to mark requires_grad using output_mask[2]?

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 3]
```diff
@@ -1023,17 +1023,12 @@ const std::vector<std::string> functions = {
                        eps : float,
                        cudnn_enabled : bool):
 
-            output, save1, save2, reserve, impl_idx = torch._batch_norm_impl_index(
-                input, weight, bias, running_mean, running_var, training,
-                momentum, eps, cudnn_enabled)
-            has_weight = weight is not None
-            has_bias = bias is not None
+            output, mean, rstd = torch.native_batch_norm(input, weight, bias, running_mean, running_var, training, momentum, eps)
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_script.cpp:1026]: This is somewhat dangerous. We don't want to disable cudnn for jitted model that's not using our fuser, there might be regression there.

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 4]
```diff
@@ -1023,17 +1023,12 @@ const std::vector<std::string> functions = {
...
 
             def backward(grad_output):
-                dinput, dweight, dbias = torch._batch_norm_impl_index_backward(
-                    impl_idx, input, grad_output, weight, running_mean, running_var,
-                    save1, save2, training, eps, [True, has_weight, has_bias], reserve)
-                return dinput, dweight, dbias, None, None, None, None, None, None
+                output_mask = [True, weight is not None, bias is not None]
+                grad_input, grad_weight, grad_bias = torch.native_batch_norm_backward(grad_output, input, weight, running_mean, running_var, mean, rstd, training, eps, output_mask)
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_script.cpp:1030]: ditto

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 5]
```diff
@@ -589,25 +589,155 @@ class IrParser {
...
+
+              TensorView* running_var = nullptr;
+              if (!node->input(4)->type()->isSubtypeOf(
+                      static_cast<c10::TypePtr>(NoneType::get()))) {
+                running_var =
+                    value_map[node->input(4)->unique()]->as<TensorView>();
+              }
+
+              // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+              // auto training = constant_as<bool>(node->input(5));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:632]: We can't ignore the training flag here, since training == false would mean different math.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 6]
```diff
@@ -589,25 +589,155 @@ class IrParser {
...
+
+              // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+              // auto momentum = constant_as<float>(node->input(6));
+              // TORCH_INTERNAL_ASSERT(
+              //     momentum.has_value(),
+              //     "The momentum (float) parameter is required.");
+              // const float kMomentum = momentum.value();
+
+              // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+              auto eps = constant_as<float>(node->input(7));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:646]: nitpick: we don't have to restrict this as constant.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 7]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
-            // const float kMomentum = momentum.value();
+            auto training = constant_as<bool>(node->input(7));
+            TORCH_INTERNAL_ASSERT(
+                training.has_value(),
+                "The training (bool) parameter is required.");
+            const bool kTraining = training.value();
 
             // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
-            auto eps = constant_as<float>(node->input(7));
+            auto eps = constant_as<float>(node->input(8));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:782]: ditto

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 8]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
-          },
-          [](const Node* node) -> bool { return true; },
-          [](const Node* node) -> OperatorType {
-            return OperatorType::Normalization;
-          });
-    }
-
-    {
-      auto ptr_op = getOperatorForLiteral(
-          "aten::layer_norm(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enable=True) -> Tensor");
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:711]: is layer_norm removed by mistake?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 9]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
-
-            // Optional: norm * weight
-            if (weight) {
-              auto weight_bcast = broadcast(weight, outer_broadcast_mask);
-              output = mul(output, weight_bcast);
+              if (output_mask[1]) {
+                auto grad_weight =
+                    sum(mul(grad_out, x_hat), outer_reduction_axes);
+                value_map.emplace(node->output(1)->unique(), grad_weight);
+              }
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:852]: github formatted this very strangely for me.
Anyway, we shouldn't leave dangling map when output_mask are false. You should still create/map an empty TensorView for the return tensor.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 10]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
+
+              auto grad_in = mul(mul(grad_out, bcast_rstd), bcast_weight);
+              value_map.emplace(node->output(0)->unique(), grad_in);
+
+              if (output_mask[1]) {
+                auto x_hat = mul(sub(input, bcast_mean), bcast_rstd);
+                auto grad_weight =
+                    sum(mul(grad_out, x_hat), outer_reduction_axes);
+                value_map.emplace(node->output(1)->unique(), grad_weight);
+              }
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:867]: ditto

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 11]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
             }
 
-            // Optional: norm * weight + bias
-            if (bias) {
-              auto bias_bcast = broadcast(bias, outer_broadcast_mask);
-              output = add(output, bias_bcast);
+            if (output_mask[2]) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:870]: ditto

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 12]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
+                  value_map[node->input(6)->unique()]->as<TensorView>();
+            }
 
             // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
-            // auto momentum = constant_as<float>(node->input(6));
-            // TORCH_INTERNAL_ASSERT(
-            //     momentum.has_value(),
-            //     "The momentum (float) parameter is required.");
-            // const float kMomentum = momentum.value();
+            auto training = constant_as<bool>(node->input(7));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:775]: If you expect this as constant, you should add a profiling rule for this. hint: `profileBool`

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 13]
```diff
@@ -620,174 +750,128 @@ class IrParser {
...
-            auto eps = constant_as<float>(node->input(7));
+            auto eps = constant_as<float>(node->input(8));
             TORCH_INTERNAL_ASSERT(
                 eps.has_value(), "The EPS parameter is required.");
             const float kEps = eps.value();
 
-            // TODO: NAN when mean and variance are zero
-            // --ftz=true -- flush-to-zero
+            // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+            auto out_mask_list = constant_as<c10::List<bool>>(node->input(9));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:788]: ditto.
This is actually a nitpick here, I think the mask would somehow always resolve as constants.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/symbolic_script.cpp:1026]: I think for forward, we can almost always safely fuse it (and mark our backend as native_batch_norm).
But for backwards, we would want to opt out of fusion if forward is running with cudnn, simply because of the opaque data structure that we can't consume.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
I'm putting a request changes here for the time being. We can discuss once you go through the comments on what changes are blocking.
I'm pretty open to merge this without fixing everything in this PR.

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp (hunk 2 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/shape_inference.cpp:239]: Bias isn't included as a parameter in `batch_norm_backward`.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 8 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/parser.cpp:711]: It was duplicated in `LayerNormFwd`.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 3 above)
Comment (rdspring1) [reply torch/csrc/jit/runtime/symbolic_script.cpp:1026]: The other versions of batch norm don't return `mean and rstd`, which we need for batch norm backward.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 5 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/parser.cpp:632]: I was planning to update this after `running_mean `and `running_var` are updating properly.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 12 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/parser.cpp:775]: How do I use `profileBool`?

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/shape_inference.cpp:239]: Sorry for not being clear, the question is, why do we want to mark requires_grad here?

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/symbolic_script.cpp:1026]: Doesn't `_batch_norm_impl_index` return all the things that we want (and even things that we don't care -> workspace :) )

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:632]: That's fair.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 8 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:711]: Got ya~ Thanks for cleaning up.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 12 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:775]: You just need to tell PE to profile certain arguments to the node as an typed data:
https://github.com/csarofeen/pytorch/blob/b08064db1e8aa1a0b431433c559449553a55992a/torch/csrc/jit/codegen/cuda/parser.cpp#L1692-L1710

graph_fuser.cpp would hook up profiling guard and convert the argument to constant for you automatically.

**Comment by csarofeen** [comment]
Closed in favor of https://github.com/csarofeen/pytorch/pull/684
